main/src/ca/uwaterloo/flix/language/phase/Terminator.scala

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2021 Matthew Lutze
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.TypedAst._
+import ca.uwaterloo.flix.language.errors.TerminationError
+import ca.uwaterloo.flix.util.Validation
+import ca.uwaterloo.flix.util.Validation.{ToFailure, ToSuccess}
+
+/**
+  * Performs checks related to termination.
+  */
+object Terminator extends Phase[Root, Root] {
+
+  override def run(root: Root)(implicit flix: Flix): Validation[Root, TerminationError] = flix.phase("Terminator") {
+    val defVal = Validation.traverseX(root.defs.values)(checkDef)
+
+    defVal map {
+      _ => root
+    }
+  }
+
+  /**
+    * Returns an error if the `defn` recurses unconditionally.
+    */
+  def checkDef(defn: Def): Validation[Unit, TerminationError] = {
+    if (unconditionallyRecurses(defn.impl.exp, defn)) {
+      TerminationError.UnconditionalRecursion(defn.sym).toFailure
+    } else {
+      ().toSuccess
+    }
+  }
+
+  /**
+    * Returns true if `exp00` unconditionally recurses, according to `recursiveAppCheck`.
+    */
+  def unconditionallyRecurses(exp00: Expression, defn: Def): Boolean = {
```

Comment:
We don't really use `exp00` anywhere. Usually it would be `exp0` and then `e0` in the inner scope.

Returning a Boolean seems pretty simple (not very future proof), but I guess its okay, because ultimately the plan is to reimplement this phase with some new technology down the road.